### PR TITLE
[ENH v2] Add partial fit to the correct branch for decisiontreeclassifier

### PR DESCRIPTION
**Pull request by adam2392**
Supersedes: #50

### CODE REVIEW

File: sklearn/tree/_tree.pyx [hunk 1]
```diff
@@ -247,6 +347,135 @@ cdef class DepthFirstTreeBuilder(TreeBuilder):
                 "lower_bound": -INFINITY,
                 "upper_bound": INFINITY,
             })
+            if rc == -1:
+                # got return code -1 - out-of-memory
+                raise MemoryError()
+
+        with nogil:
+            while not update_stack.empty():
```
Comment (adam2392) [inline sklearn/tree/_tree.pyx:355]: @PSSF23 does this code affect the normal building process not via `partial_fit`

Comment (PSSF23) [reply sklearn/tree/_tree.pyx:355]: it shouldn't. update_stack only exists when the initial_roots has content.

**Review by PSSF23 — COMMENTED** [review]
@adam2392 Yes, the `classes` param was not properly passed in by the CI, so it was expecting a different kind of error. We can remove the `_check_partial_fit_first_call` check, but then there might be other errors associated with `partial_fit` api.

By the way, the performance is improved with the new branch, about the same as streaming tree on its own now. So previously the inferior results might be due to some bug in `submodulev2`.

**Comment by PSSF23** [comment]
It appears that they are testing all methods with fitting functions (`fit`, `partial_fit`, ...) about `ccp_alpha`, but did not give `partial_fit` the `classes` variable it requires. As long as `_check_partial_fit_first_call` exists there will be errors, because `partial_fit` could not run properly without `classes`. It was added because it was a requirement for the `partial_fit` api.

The checks for `ccp_alpha`, which they are testing, exist in `_build_pruned_tree_ccp`. So it can only be tested correctly after there's a tree.

**Review by PSSF23 — COMMENTED** [review]
For unit tests, `sklearn` had built-in checks for `partial_fit` from other streaming algorithms, so I didn't prepare any extra ones. We can add new ones as we see fit, but the existing ones should cover most situations.

**Comment by adam2392** [comment]
> For unit tests, `sklearn` had built-in checks for `partial_fit` from other streaming algorithms, so I didn't prepare any extra ones. We can add new ones as we see fit, but the existing ones should cover most situations.

I see. I took a look and all you have to do is validate the input parameters before running the partial fit check: https://github.com/neurodata/scikit-learn/pull/54/commits/63637f7da3469195e04339dc0a9c5aeae99407ed

**Comment by adam2392** [comment]
Do you want to add the support for `partial_fit` into `ForestClassifier`?

**Review by PSSF23 — APPROVED** [review]
@adam2392 Thanks for all the improvements. I need to work on simulations for the cancer project this week so feel free to go ahead. We also found some potential bugs on honest code so I might need to revisit it.

**Comment by adam2392** [comment]
Kay merged! @PSSF23 feel free to open up a PR to enable this for the `ForestClassifier` class.

**Pull request merged**